File: agent-service/app/project/inner_prompts/project_index.py
"""项目总索引 Prompt。"""
from __future__ import annotations

from enum import Enum

__all__ = ["ProjectIndexPrompt"]


class ProjectIndexPrompt(str, Enum):
    """项目总索引生成 Prompt 模板。"""

    PROJECT_INDEX = """
你是 PM-Agent 的项目总索引生成器，需要根据文件树扫描结果、文件详情摘要和已有 `index.json`，生成符合 `index.json` 规范的轻量路由索引。

# 输入
你会收到以下内容中的一种或多种：
1. `existing_index_json`：已有 `index.json`，可能为空。
2. `scan_result_json`：文件树扫描结果，包括 path、name、size_bytes、quick_fingerprint、content_hash、scan_status 等。
3. `file_detail_summaries`：已生成的 `file_details` 摘要列表。
4. `storage_meta`：存储信息，包括 provider、bucket、business、object_prefix、logical_root。
5. `source_meta`：来源信息，例如 project_id、generated_at。

# 职责边界
你只负责生成轻量总索引。
禁止输出原始文件全文、大段代码、大段文档、Prompt 内容、内部评分细节和敏感信息。
禁止把 `file_details.content_slices` 完整复制进 `index.json`。

# 生成规则
1. `entries[].id` 必须稳定，文件移动或内容变化时保留原 id。
2. `entries[].path` 使用源文件当前相对路径。
3. `entries[].detail_ref` 使用逻辑路径，例如 `file_details/backend/auth/file-backend-auth-controller.json`。
4. `quick_fingerprint` 使用已计算后的字符串，不输出 size、mtime 等计算材料。
5. `content_hash` 用于精确判断内容变化，必须来自扫描结果或文件详情。
6. `keywords` 建议 3 到 12 个。
7. `summary` 建议不超过 80 个中文字符。
8. 被忽略文件可进入统计摘要，不进入 `entries`。
9. 删除文件保留 entry，并设置 `status=deleted`。

# kind 枚举建议
- `backend_code`
- `frontend_code`
- `agent_code`
- `config`
- `doc`
- `test`
- `script`
- `migration`
- `unknown`

# 输出要求
只输出合法 JSON，不要输出 Markdown、解释文本或代码块。
字段名必须使用英文，字段值中的自然语言内容使用中文。

# JSON 格式
{
  "project_id": "项目 ID",
  "schema_version": "1.0.0",
  "generated_at": "2026-07-02T00:00:00",
  "storage": {
    "provider": "minio",
    "bucket": "pm-agent",
    "business": "project",
    "object_prefix": "PM-AGENT/u001/local-pm-agent-001/project/context/",
    "logical_root": "context/"
  },
  "summary": {
    "total_nodes": 0,
    "active_files": 0,
    "ignored_files": 0,
    "deleted_files": 0,
    "detail_files": 0
  },
  "entries": [
    {
      "id": "稳定文件 ID",
      "path": "源文件相对路径",
      "module": "粗粒度业务模块",
      "kind": "backend_code | frontend_code | agent_code | config | doc | test | script | migration | unknown",
      "language": "java | typescript | vue | python | markdown | json | yaml | other",
      "status": "active | deleted | ignored | pending_review",
      "importance": "high | medium | low",
      "quick_fingerprint": "qf:sha256:xxx",
      "content_hash": "sha256:xxx",
      "summary": "不超过 80 个中文字符的职责摘要",
      "keywords": ["关键词"],
      "detail_ref": "file_details/backend/auth/file-backend-auth-controller.json",
      "updated_at": "2026-07-02T00:00:00"
    }
  ],
  "refs": {
    "file_details": "file_details/",
    "project_specification": "project_specification.json",
    "long_term_memory": "long_term_memory.json",
    "short_term_memory": "short_term_memory.json",
    "user_habits": "user_habits/",
    "update_journal": "update_journal.jsonl"
  }
}
""".strip()

File: backend/src/main/java/com/ning/pm/project/context/dto/ProjectIndex.java
package com.ning.pm.project.context.dto;

import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;

import java.time.LocalDateTime;
import java.util.List;

/**
 * ProjectIndex 表示写入 system/index.json 的完整项目上下文索引。
 *
 * @author ning
 * @date 2026-07-15
 */
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
public record ProjectIndex(
        Long projectId,
        String projectName,
        Long ownerUserId,
        String schemaVersion,
        LocalDateTime generatedAt,
        LocalDateTime updatedAt,
        Storage storage,
        Summary summary,
        List<FileEntry> project,
        List<FileEntry> user,
        List<UploadFailure> uploadFailures,
        SystemSection system
) {

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record Storage(
            String provider,
            String bucket,
            String objectPrefix,
            String indexPath
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record Summary(
            long totalNodes,
            long activeFiles,
            long failNodes
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record FileEntry(
            Long id,
            String storageUuid,
            String logicalPath,
            String fileName,
            String storageName,
            String minioPath,
            Long sizeBytes,
            String contentType,
            String status,
            String analysisStatus,
            String quickFingerprint,
            String contentHash,
            String module,
            String kind,
            String language,
            String importance,
            String summary,
            List<String> keywords,
            String detailRef,
            LocalDateTime updatedAt
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record UploadFailure(
            Long fileId,
            String business,
            String logicalPath,
            String fileName,
            String storageName,
            String status,
            Integer attempts,
            String lastErrorCode,
            LocalDateTime updatedAt
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record SystemSection(
            String index,
            String fileDetails,
            String projectSpecification,
            String longTermMemory,
            String shortTermMemory,
            String userHabits,
            String updateJournal
    ) {
    }
}
